builtin/logical/database/secret_creds.go

```diff
@@ -0,0 +1,129 @@
+package database
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+)
+
+const SecretCredsType = "creds"
+
+func secretCreds(b *databaseBackend) *framework.Secret {
+	return &framework.Secret{
+		Type:   SecretCredsType,
+		Fields: map[string]*framework.FieldSchema{},
+
+		Renew:  b.secretCredsRenew(),
+		Revoke: b.secretCredsRevoke(),
+	}
+}
+
+func (b *databaseBackend) secretCredsRenew() framework.OperationFunc {
+	return func(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+		// Get the username from the internal data
+		usernameRaw, ok := req.Secret.InternalData["username"]
+		if !ok {
+			return nil, fmt.Errorf("secret is missing username internal data")
+		}
+		username, ok := usernameRaw.(string)
+
+		roleNameRaw, ok := req.Secret.InternalData["role"]
+		if !ok {
+			return nil, fmt.Errorf("could not find role with name: %s", req.Secret.InternalData["role"])
+		}
+
+		role, err := b.Role(req.Storage, roleNameRaw.(string))
+		if err != nil {
+			return nil, err
+		}
+		if role == nil {
+			return nil, fmt.Errorf("could not find role with name: %s", req.Secret.InternalData["role"])
```

Comment:
This should be a (at least slightly) different message from up above just to make it easier, if someone hits this error, to pinpoint exactly where it came from.